File: python/resting-state.py
from psychopy import core, visual, event, monitors, parallel, prefs
from psychopy import sound
# prefs.general['audioLib'] = ['pygame']


# Resting State Script
# Jan Freyberg (jan.freyberg@kcl.ac.uk)
# This script simply presents instructions, and sends triggers.

# Parameters for this script
trialnumber = 4  # how many trials there will be in total
trialduration = 60  # how long a trial is

# Triggers - the script will just cycle through these
# i.e. for 2 triggers & 4 trials, pattern will be A B A B
# make sure trialnumber is a multiple of len(triggers) for balanced design
triggers = [101, 102]

# Set the screen parameters: (This is important!)
screen = monitors.Monitor('tobiix300')
screen.setSizePix([1920, 1080])  # resolution of screen in pixels
screen.setWidth(51)  # screen width in cm
screen.setDistance(60)  # distance from screen in cm

# Open the display window:
win = visual.Window([500, 500], allowGUI=True, monitor=screen,
                    units='deg', fullscr=True)

# Make a generic text stimulus for instructions
message = visual.TextStim(win, units='norm', pos=[0, 0], height=0.07,
                          alignVert='center', alignHoriz='center',
                          text='')

# Make a fixation cross
fixation = visual.GratingStim(win, tex='sqr', mask='cross', sf=0, size=0.3,
                              pos=[0, 0], color='black', autoDraw=False)

# Make a sound that will notify the user when the trial changes
beep = sound.Sound(value=400, secs=0.25)
beep.setVolume(0.4)


# define a trigger function. change this depending on the system
def trigger(value=0):
    # send trigger
    print(value)


# Define a function that takes a string and displays it, then waits to proceed
def instruct(displaystring):
    message.text = displaystring
    message.draw()
    win.flip()
    event.waitKeys(keyList=['space'])

# Give instructions:
instruct("Please relax for the next 4 minutes. Please alternate between "
         "keeping your eyes open for a minute and keeping your eyes "
         "closed for a minute (you can of course blink.) There will be "
         "a tone notifying you of when you need to switch from having "
         "your eyes open to closed, or vice versa.\n\n"
         "Press [space] to continue.")
instruct("When you hear the first tone, please start with your eyes "
         "open for one minute. Please try to focus your eyes on the "
         "fixation cross in the center of the screen.\n\nIf you have any "
         "questions, please ask now - if not, check if the experimenter "
         "is ready, and then press [space] to start.")

for seconds in range(5):
    message.text = ("Starting in " + str(5 - seconds) + " seconds.")
    message.draw()
    win.flip()
    core.wait(1)
win.flip()

# Some stuff to keep track of timing
trialClock = core.Clock()
trialClock.add(1)  # because we want to have time before the trial starts
# Go through trials!
for trials in range(trialnumber):
    # before the trial is supposed to start, wait with high precision
    while trialClock.getTime() < trials * trialduration:
        pass
    beep.play()  # announce the next trial
    trigger(triggers[trials % len(triggers)])  # trigger for eeg
    trialClock.reset()  # reset clock
    win.flip()  # flip (no real reason)
    # step through trial duration (except for one second) in 1s chunks
    for seconds in range(int(trialduration - 1)):
        core.wait(1)
        # check if experiment was interrupted in last second
        if event.getKeys(keyList='escape'):
            trigger(99)
            raise KeyboardInterrupt("You interrupted the script manually!")

File: matlab/resting_state.m
%% A Resting State Script

% Variables
trialnumber = 4; % how many trials to run through
trialduration = 60; % trial duration in seconds

% Triggers
triggers = [101, 102]; % what triggers you want to use
triggers = repmat(triggers, 1, trialnumber/numel(triggers));

% Backgroundcolor
bgcolor = 127.5;

try
    InitializePsychSound;
    pa = PsychPortAudio('Open', [], [], [], [], [], 256);
    bp400 = PsychPortAudio('CreateBuffer', pa, [MakeBeep(400, 0.2); MakeBeep(400, 0.2)]);
    PsychPortAudio('FillBuffer', pa, bp400);
catch sounderr
    sca;
    rethrow(sounderr);
end

% Configure the Parallel Port
try
    % config_io;
    address = hex2dec('D010');
    outp(address, 0); % set all parport pins low
catch ioerr
    sca;
    disp('IO Error. Is config_io installed and is it the right version?');
    rethrow(ioerr);
end

KbName('UnifyKeyNames');
esc_key = KbName('Escape');

% Open the screen
try
    scr = Screen('OpenWindow', 2, bgcolor);
    HideCursor;
catch screenerr
    sca;
    rethrow(screenerr);
end

try
    % Instruct the participant:
    DrawFormattedText(scr, ['From the first beep, keep your eyes open for a minute.'...
                            'After a minute you will hear a second beep, indicating '...
                            'you should close your eyes for one minute. This will '...
                            'continue to alternate for 4 minutes.\n\nPlease try and '...
                            'restrict blinking during the eye open intervals. Press '...
                            'any key to start.'], 'center', 'center', 0, 40, [], [], 1.5);
    Screen('Flip', scr);
    KbStrokeWait;

    % Give a 5 Second countdown:
    for lapsedTime = 1:5
    DrawFormattedText(scr, ['Starting in ' num2str(6-lapsedTime) ' seconds'], 'center', 'center', 0);
    Screen('Flip', scr);
    WaitSecs(1);
    end

    % Draw the fixation cross
    DrawFormattedText(scr, '+', 'center', 'center', 0);
    Screen('Flip', scr);

    % Priority(1); % NB Priority works differently on Windows, OSX and Linux
    %              % Priority(1) is highest priority on Windows machines


    % Cycle through trials:
    t = GetSecs + 1; % Add one sec to give an additional 1 sec waiting
    for trial = 1:trialnumber
        % Before the trial, wait with high precision
        while GetSecs < t
            WaitSecs(0.001); % Wait one millisecond
        end
        % Start the trial:
        outp(address, triggers(trial)); % Send the trigger
        PsychPortAudio('Start', pa); % Produce the beep

        t = GetSecs + trialduration; % Set start time for next trial

        % Wait for the next trial by waiting in 1second steps:
        for waitTime = 1:trialduration-5
            % Wait one sec & also check keyboard:
            [~, keys] = KbWait([], [], GetSecs+1);
            % If escape key was pressed:
            if keys(esc_key)
                error('You manually interrupted the script.');
            end
        end
    end

catch experimenterr
    sca;
    rethrow(experimenterr)
end

% Finish: Beep twice, close everything
PsychPortAudio('Start', pa);
WaitSecs(0.5);
PsychPortAudio('Start', pa);
sca;
